src/ensadd.fd/ENSADD.f90

```diff
@@ -0,0 +1,187 @@
+!
+!   OUTPUT FILES:
+!     UNIT  51  GRIB FILE 
+!
+!   SUBPROGRAMS CALLED:
+!     GETGB2 -- W3LIB ROUTINE
+!     PUTGB2 -- W3LIB ROUTINE
+!
+! ATTRIBUTES:
+!   LANGUAGE: FORTRAN
+!
+!$$$
+
+program      ensadd_g2
+
+use grib_mod
+use params
+
+implicit none
+
+type(gribfield) :: gfldo
+
+integer :: currlen=0
+logical :: unpack=.true.
+logical :: expand=.false.
+
+integer,dimension(200) :: kids,kpdt,kgdt
+integer kskp,kdisc,kpdtn,kgdtn,i
+
+integer,dimension(200) :: jids,jpdt,jgdt,iids,ipdt,igdt
+integer jskp,jdisc,jpdtn,jgdtn,idisc,ipdtn,igdtn
+
+integer temp(200)
+
+integer ienst,iensi,lpgb,lpgi,lpge,icount,iretb,ireti,irete,iret,jret,ipdtnum_out                                        
+character*255 cpgb,cpge
+namelist     /namin/ ienst,iensi,cpgb,cpge      
+
+CALL W3TAGB('ENSADD',1999,0243,0068,'NP23')
+
+read (5,namin)
+lpgb=len_trim(cpgb)
+lpge=len_trim(cpge)
+
+print *, cpgb(1:lpgb),' ',cpge(1:lpge)
+print *, ' '  
+
+call baopenr(11,cpgb(1:lpgb),iretb)
+call baopen (51,cpge(1:lpge),irete)
+
+! loop over variables
+
+kids=-9999;kpdt=-9999; kgdt=-9999
+kdisc=-1;  kpdtn=-1;   kgdtn=-1
+
+icount=0
+kskp=0
+do
+
+  print *, '----- Read Ensemble Forecast ------'
+  print *, ' '  
+
+  call getgb2(11,0,kskp,kdisc,kids,kpdtn,kpdt,kgdtn,kgdt,unpack,kskp,gfldo,iret)
+
+  if(iret.ne.0) then
+    if(iret.eq.99 ) exit
+    print *,' getgb2 error = ',iret
+    cycle
+    !call errexit(17)
+  endif
+
+  icount=icount+1
+
+  if(iret.eq.0) then
+    call printinfr(gfldo,icount)
+  else
+    print*, 'there is no fcst for ens member'
+  endif
+
+  ! gfldo%idsect(2): Identification of originating
+  ! gfldo%idsect(2)=2: NCEP Ensemble Products
+  ! Original GFS has gfldo%idsect(2)=0, change it to be 2
+
+  gfldo%idsect(2)=2
+  gfldo%idsect(13)=3
+
+  ! when product difinition template 4.0 change to 4.1
+  ! when product difinition template 4.8 change to 4.11
+  ! ipdtlen aslo change, need do modification for output
+
+! print*, '1 gfldo%ipdtmpl=',gfldo%ipdtmpl   
+
+  if(gfldo%ipdtnum.eq.0) then
+
+    temp=-9999
+
+    temp(1:gfldo%ipdtlen)=gfldo%ipdtmpl(1:gfldo%ipdtlen)
+    if(gfldo%ipdtmpl(3).eq.2) then
+     temp(3)=4
+    endif 
+
+    deallocate (gfldo%ipdtmpl)
+
+    gfldo%ipdtnum=1
+    if(gfldo%ipdtnum.eq.1) gfldo%ipdtlen=18
+    if(gfldo%ipdtnum.eq.1) allocate (gfldo%ipdtmpl(gfldo%ipdtlen))
+
+    gfldo%ipdtmpl(1:15)=temp(1:15)
+    gfldo%ipdtmpl(16)=ienst
+    gfldo%ipdtmpl(17)=iensi
+    gfldo%ipdtmpl(18)=10    
+
+  endif
+
+  if(gfldo%ipdtnum.eq.8) then
+
+    temp=-9999
+
+    temp(1:gfldo%ipdtlen)=gfldo%ipdtmpl(1:gfldo%ipdtlen)
+    if(gfldo%ipdtmpl(3).eq.2) then
+     temp(3)=4
+    endif 
+
+    deallocate (gfldo%ipdtmpl)
+
+    gfldo%ipdtnum=11
+    if(gfldo%ipdtnum.eq.11) gfldo%ipdtlen=32
+    if(gfldo%ipdtnum.eq.11) allocate (gfldo%ipdtmpl(gfldo%ipdtlen))
+
+    gfldo%ipdtmpl(1:15)=temp(1:15)
+    gfldo%ipdtmpl(16)=ienst
+    gfldo%ipdtmpl(17)=iensi
+    gfldo%ipdtmpl(18)=10    
```

Comment:
Please modify line 152 from
gfldo%ipdtmpl(18)=10
to
gfldo%ipdtmpl(18)=30

Reply:
Is it that wrong in ops too then?

This should be turned into a parameter that is passed in instead of a hard-coded value. Right now this PR was just to get the GEFS code copied over to gfs-utils. Modifications can come later.

Reply:
I  reviewed ush/gefs_prdgen.sh and checked why the ush/global_ensadd.sh is needed.  It seems that global_ensadd.sh is used for GFS file only. The GFS forecast is thought to be one member of GEFS and has perturbation numbers ( 0 0 )

# Add ensemble PDS header to GFS file
                $ENSADD 0 0 pgb2afile.$ffhr epgbafile

Please correct me if I am wrong. For such a case, there is no need to modify the code.